### PR TITLE
[linker] Add metadata to linked assemblies indicating the RID for which they were created

### PR DESCRIPTION
**Pull request by grendello**
As a preparation for future effort to properly support per-RID assemblies, this PR
adds a linker step which appends the `AssemblyMetadata` custom attribute to each
assembly, with the key `XamarinAndroidAbi` and its value set to the RID for which
the assembly was linked.

Right now this metadata is not used, but in the future we plan to put **all** the
assemblies in per-architecture shared libraries, without any exceptions.  This will
be done to ensure that unknown changes made to assemblies by the linker will **not**
break the application at runtime (since the runtime will not verify and complain
about "platform incompatible" assemblies, it doesn't have any notion of such
assemblies)

### CODE REVIEW

**Comment by grendello** [comment]
Converted to draft, because the test failures cannot be fixed without the planned changes to package all assemblies for all RIDs in shared libraries, work that hasn't been started yet.

I'm going to keep this PR open and updated as draft, until the other changes are implemented.

File: src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.ILLink.targets [hunk 1]
```diff
@@ -40,6 +40,7 @@ This file contains the .NET 5-specific targets to customize ILLink
         https://github.com/dotnet/sdk/blob/a5393731b5b7b225692fff121f747fbbc9e8b140/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets#L147
         -->
       <_TrimmerCustomData Include="XATargetFrameworkDirectories" Value="$(_XATargetFrameworkDirectories)" />
+      <_TrimmerCustomData Include="XARuntimeIdentifier" Value="$(RuntimeIdentifier)" />
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.ILLink.targets:43]:
```suggestion
      <_TrimmerCustomData Include="XARuntimeIdentifier" Value="$(RuntimeIdentifier)"  Condition=" '$(_AndroidAddRuntimeIdentifierToAssemblies)' == 'true' " />
```

File: src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/AddRidMetadataAttributeStep.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		if (String.IsNullOrEmpty (rid)) {
+			throw new InvalidOperationException ("RID must have a non-empty value");
+		}
+
+		AssemblyDefinition corlib = GetCorlib ();
+		MethodDefinition assemblyMetadataAttributeCtor = FindAssemblyMetadataAttributeCtor (corlib);
+		TypeDefinition systemString = GetSystemString (corlib);
+
+		var attr = new CustomAttribute (assembly.MainModule.ImportReference (assemblyMetadataAttributeCtor));
+		attr.ConstructorArguments.Add (new CustomAttributeArgument (systemString, "XamarinAndroidAbi")); // key
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/AddRidMetadataAttributeStep.cs:41]: Should we avoid naming things with an `XA` or `Xamarin` prefix? Can this just be `RuntimeIdentifier`?

Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/AddRidMetadataAttributeStep.cs:41]: `RuntimeIdentifier` is too generic IMO. `XA` or `XamarinAndroid` makes it less likely to clash with other people's metadata

**Comment by grendello** [comment]
With the change from @jonathanpeppers it can actually go in

File: src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/AddRidMetadataAttributeStep.cs (hunk 2 above)
Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/AddRidMetadataAttributeStep.cs:41]: Or `AndroidRuntimeIdentifier`? I don't think they want us to introduce new things that say `Xamarin`.

Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/AddRidMetadataAttributeStep.cs:41]: Who would know? :)

**Pull request closed without merging**